File: POH7/SantaFuku/Main.clj
; Try POH
; author: Leonardone @ NEETSDKASU
(def gs read-line)
(defn gi [] (read-string (gs)))
(defn gss [] (list* (.split (gs) " "))) ; gisのmapがlistを返すのでgssも揃える
(defn gis [] (map read-string (gss)))
(defn ngt [n f] (reverse (loop [c n r []] (if (= c 0) r (recur (dec c) (cons (f) r))))))  ; forやmapだと遅延評価で読み込み順が狂うう
(defn ngs [n] (ngt n gs))
(defn ngi [n] (ngt n gi))
(defn ngss [n] (ngt n gss))
(defn ngis [n] (ngt n gis))
; ============================================

(defn minval [vs v]
    (let [vss (sort vs)]
        (first (reduce (fn [[r b] e] [(min r (- e b)) e]) [v 0] vss))
    ))
(defn solve [x y z da]
    (let [buildf (fn [[xs ys] [d a]] (if (= d 0) [(cons a xs) ys] [xs (cons a ys)]))
          [xs ys] (reduce buildf [[x] [y]] da)
          xmin (minval xs x)
          ymin (minval ys y)
          ]
        (* xmin ymin z)
    ))

; ============================================

(defn main []
    (let [[x y z n] (gis)
          da (ngis n)
          ans (solve x y z da)]
        (println ans)
    ))

; ============================================
(main)
; ============================================

File: POH7/SantaFuku/Main.pl
# Try POH
# author: Leonardone @ NEETSDKASU
# -----------------------------------------------------------------------------
sub gs { chomp(my $s = <>); $s }
sub gi { int gs }
sub gss { split / /, gs  }
sub gis { map int, gss }
sub ngt { my ($n, $f) = @_; my @r = (); for (1 .. $n) { push @r, &$f } @r }
sub ngs { ngt shift @_, \&gs }
sub ngi { ngt shift @_, \&gi }
sub ngss { ngt shift @_, sub { [gss] } }
sub ngis { ngt shift @_, sub { [gis] } }
sub min { my $v = pop @_; for $x (@_) { if ($x < $v) { $v = $x} } $v }
sub max { my $v = pop @_; for $x (@_) { if ($x > $v) { $v = $x} } $v }
sub isort { sort {$a <=> $b} @_ }
# -----------------------------------------------------------------------------


sub solve {
    my ($x, $y, $z, $n) = gis;
    my @xs = (0, $x);
    my @ys = (0, $y);
    for (ngis $n) {
        my ($d, $a) = @{$_};
        if ($d == 0) {
            push @xs, $a;
        } else {
            push @ys, $a;
        }
    }
    sub minval {
        my $v = shift @_;
        my $p = shift @_;
        for $w (@_) {
            $v = min $v, $w - $p;
            $p = $w
        }
        $v;
    }
    $x = minval($x, isort @xs);
    $y = minval($y, isort @ys);
    my $ans = $x * $y * $z;
    print "$ans\n";
}

solve;
